src/dtls.c

```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                if (found) {
+                    pskInfo->cipherSuite0 = foundSuite[0];
+                    pskInfo->cipherSuite  = foundSuite[1];
+                    pskInfo->isValid      = 1;
+                    return;
+                }
+            }
+        }
+    }
+}
+#endif
+
+#ifdef WOLFSSL_DTLS13
+
+#ifndef WOLFSSL_SEND_HRR_COOKIE
+#error "WOLFSSL_SEND_HRR_COOKIE has to be defined to use DTLS 1.3 server"
+#endif
+
+static int SendStatelessReplyDtls13(const WOLFSSL* ssl, WolfSSL_CH* ch,
+                                    PskInfo* pskInfo)
+{
+    int ret = -1;
+
+    (void)pskInfo;
+
+    if (ch->cookie.size == 0) {
+        TLSX* parsedExts = NULL;
+        WolfSSL_ConstVector tlsx;
+        Suites suites;
+        word16 len;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+        byte haveKS = 0;
+        byte usePSK = 0;
+        byte doKE = 0;
+#endif
+        CipherSuite cs;
+        CipherSpecs specs;
+        byte cookieHash[WC_MAX_DIGEST_SIZE];
+        int cookieHashSz;
+
+        XMEMSET(&cs, 0, sizeof(cs));
+
+        /* We need to echo the session ID sent by the client */
+        if (ch->sessionId.size > ID_LEN) {
+            /* Too large. We can't echo this. */
+            ERROR_OUT(INVALID_PARAMETER, dtls13_cleanup);
+        }
+
+        /* Populate the suites struct to find a common ciphersuite */
+        XMEMSET(&suites, 0, sizeof(suites));
+        suites.suiteSz = (word16)ch->cipherSuite.size;
+        if ((suites.suiteSz % 2) != 0)
+            ERROR_OUT(INVALID_PARAMETER, dtls13_cleanup);
+        if (suites.suiteSz > WOLFSSL_MAX_SUITE_SZ)
+            ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+        XMEMCPY(suites.suites, ch->cipherSuite.elements, suites.suiteSz);
+
+        /* Populate extensions */
+
+        /* Supported versions always need to be present. Has to appear after
+         * key share as that is the order we reconstruct it in
+         * RestartHandshakeHashWithCookie. */
+        ret = TLSX_Push(&parsedExts,
+                  TLSX_SUPPORTED_VERSIONS, ssl, ssl->heap);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        /* Set that this is a response extension */
+        parsedExts->resp = 1;
+
+        ret = CopyExtensions(ssl->extensions, &parsedExts, ssl->heap);
+        if (ret != 0)
+            goto dtls13_cleanup;
+
+        /* Signature algs */
+        ret = TlsxFindByType(&tlsx, TLSX_SIGNATURE_ALGORITHMS,
+                             ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size > OPAQUE16_LEN) {
+            ato16(tlsx.elements, &len);
+            if (len != tlsx.size - OPAQUE16_LEN)
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            if ((len % 2) != 0)
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            suites.hashSigAlgoSz = len;
+            XMEMCPY(suites.hashSigAlgo, tlsx.elements + OPAQUE16_LEN,
+                    len);
+        }
+
```

Comment:
Is `SIGNATURE_ALGORITHMS` extension obligatory if not using PSK? If so, should we error out?

Reply:
Adding check and erroring out before `MatchSuite_ex`.